velox/functions/sparksql/DecimalArithmetic.cpp

```diff
@@ -14,498 +14,358 @@
  * limitations under the License.
  */
 
-#include "velox/common/base/CheckedArithmetic.h"
-#include "velox/expression/DecodedArgs.h"
-#include "velox/expression/VectorFunction.h"
+#include "velox/functions/Macros.h"
+#include "velox/functions/Registerer.h"
 #include "velox/functions/sparksql/DecimalUtil.h"
-#include "velox/type/DecimalUtil.h"
 
 namespace facebook::velox::functions::sparksql {
 namespace {
 
-std::string getResultScale(std::string precision, std::string scale) {
-  return fmt::format(
-      "({}) <= 38 ? ({}) : max(({}) - ({}) + 38, min(({}), 6))",
-      precision,
-      scale,
-      scale,
-      precision,
-      scale);
-}
-
-// Returns the whole and fraction parts of a decimal value.
-template <typename T>
-inline std::pair<T, T> getWholeAndFraction(T value, uint8_t scale) {
-  const auto scaleFactor = velox::DecimalUtil::kPowersOfTen[scale];
-  const T whole = value / scaleFactor;
-  return {whole, value - whole * scaleFactor};
-}
+struct BinaryPrecisionScales {
+  uint8_t aPrecision;
+  uint8_t aScale;
+  uint8_t bPrecision;
+  uint8_t bScale;
+};
 
-// Increases the scale of input value by 'delta'. Returns the input value if
-// delta is not positive.
-inline int128_t increaseScale(int128_t in, int16_t delta) {
-  // No need to consider overflow as 'delta == higher scale - input scale', so
-  // the scaled value will not exceed the maximum of long decimal.
-  return delta <= 0 ? in : in * velox::DecimalUtil::kPowersOfTen[delta];
+// Extracts the precisions and scales for two decimal types.
+BinaryPrecisionScales getDecimalPrecisionScales(
```

Comment:
I'm wondering about the benefits of BinaryPrecisionScales struct and this function. It seems using getDecimalPrecisionScale directly would be clearer. Am I missing something? CC: @rui-mo

Reply:
This function is added to reuse the process of extracting precision and scales. It's fine to use getDecimalPrecisionScale directly if you think that is clearer. Thanks.

Reply:
I think it will be cleaner to just use 2 getDecimalPrecisionScale calls.